docs(wysiwyg): Added documentations and guides on how to use Wysiwyg (#896)
front-commerce/developers.front-commerce.com: docs/02-guides/customize-wysiwyg-platform.mdx

---
sidebar_position: 4
hide_table_of_contents: false
title: Customize WYSIWYG Platform
description:
  In this guide, you will learn the different WYSIWYG implementations available
  on Front-Commerce and how to customize them.
---

import TOCInline from "@theme/TOCInline";
import ContactLink from "@site/src/components/ContactLink";

> import SinceVersion from "@site/src/components/SinceVersion";

Reviewer: No need to import these, they are added to the component library: https://github.com/front-commerce/developers.front-commerce.com/blob/main/src/theme/MDXComponents.ts

<SinceVersion tag="3.4" />

<p>{frontMatter.description}</p>

If you want to learn how the core WYSIWYG component works instead, please refer
to [Display WYSIWYG content](/docs/3.x/guides/display-wysiwyg-content).

> Each platform has a specific type of WYSIWYG. This allows to render your content
> differently depending of its origin. For instance a content from WordPress might
> have some specific media shortcodes while Magento will have some widgets to
> display a category name. In the following section you will learn about the one
> implemented in Front-Commerce:

Reviewer: ```suggestion
Each platform has a specific type of WYSIWYG. This allows to change how your content is rendered depending on its origin. For instance, a content from WordPress might
have some specific media shortcodes while Magento will have some widgets to
display a category name. In the following section you will learn about the one
implemented in Front-Commerce:
```

<div className="pl-4">
  <TOCInline toc={toc} maxHeadingLevel={4} />
</div>

Definitions:

- a _shortcode_ is a specific string structure that is meant to be transformed
  into actual content

> - a _transform_ is function that replaces an HTML tag with a React Component

Reviewer: ```suggestion
- a _transform_ is a function that replaces an HTML tag with a React Component
```

  components when the `href` attribute does not contain a domain.
- `<widget>` tags are transformed into
  `theme/modules/Wysiwyg/MagentoWysiwyg/Widget/Widget.js` components. However,
  you shouldn't write a `<widget>` tag manually. It comes from the
  `{% raw %}{{widget}}{% endraw %}` shortcode.
- `<style>` tags are transformed to support
  [Magento's Page Builder](/docs/3.x/extensions/magento2/how-to/use-and-extend-the-page-builder)
  format. The `#html-body` selector is replaced with the `<Wysiwyg>` root
  selector. See
  [WYSIWYG dynamic styles](/docs/3.x/guides/display-wysiwyg-content#dynamic-styles)
  for details.

### Add a custom Magento Widget

Custom widgets will be automatically parsed. However, you will still need to map
the widget's `type` to custom React Components. If you don't, they will be
ignored and nothing will be rendered in the final page.

To do so, please override
`theme/modules/Wysiwyg/MagentoWysiwyg/Widget/getWidgetComponent.js` in your own
theme.

Now, within your newly created `getWidgetComponent.js` file, you will be able to
add your own behavior. For instance, if you've created a widget in Magento that

> uses the type `acme/product-preview`, you will need to update the `Widget.js`
> file likewise:

Reviewer: ```suggestion
uses the type `acme/product-preview`, you will need to update the `Widget.js`
file:
```

```js title="theme/modules/Wysiwyg/MagentoWysiwyg/Widget/getWidgetComponent.js"
import React from "react";
import Spinner from "theme/components/atoms/Spinner";
import loadable from "@loadable/component";

const defaultWidgetsMap = {
  "acme/product-preview": loadable(
    () => import("./path/to/ProductPreview.js"),
    {
      fallback: <Spinner />,
    }
  ),

> };

Reviewer: We no longer support loadable, this should be replaced with react `lazy()` and `Suspense`

```

If you restart your application, you will notice that the new widget component
is displayed. Hurray!

However, in most cases, you will need to fetch data to display all the needed
information in your widget. For instance, if the widget is
`{% raw %}{{ widget type="acme/product-preview" sku="VSK12" }}{% endraw %}` you
will want to fetch the product associated with the given SKU.

#### Register your widget type at the GraphQL level

1.  First make sure that your dependencies are up to date in your GraphQL module
    ```ts title="extensions/acme-extension/modules/wysiwyg/index.ts"
    export default createGraphQLModule({
    namespace: "Magento2/Cms",
    // highlight-start
    dependencies: [
      "Magento2/Wysiwyg", // ensure that Widget related features are available
      "Magento2/Catalog/Product" // ensure that you can fetch a product in your Wysiwyg data
    ],
    // highlight-end
    ```
1.  Then you need to register a new WidgetData type in your schema
    ```ts title="extensions/acme-extension/modules/wysiwyg/index.ts"
    typeDefs: /* GraphQL */ `
      type WidgetProductPreviewData implements WidgetData {
        dataId: ID
        product: Product
      }

>     `,

Reviewer: we now inline the `typeDefs` it should be in the future ADR for https://linear.app/front-commerce/issue/FC-2044/[vite]-refactor-graphql-modules

    ```
1.  After that you can setup the resolvers to tell GraphQL how to fetch the
    `product` field
    ```ts title="extensions/acme-extension/modules/wysiwyg/resolvers.ts"
    export default {
      WidgetProductPreviewData: {
        product: ({ node }, _, { loaders }) => {
          const productSkuAttribute = node.attrs.find(
            ({ name }) => name === "sku"
          );
          if (!productSkuAttribute) {
            return null;
          }
          return loaders.Product.load(productSkuAttribute.value);
        },
      },

>     };

Author: I agree for typeDefs, but that is not true for resolvers. In fact, we still have `resolvers.js` files for most, if not all of our graphql modules.

I believe inlining resolvers will end up bloating most of `runtime.ts` files.

>     };

Reviewer: Yes although we still have them, we will include it in the ADR that best practice is to inline them, this is so that they can benefit from the types of the `createGraphQLRuntime()` which eventually will include the context param types for all loaders, including config, user, services, etc..

>     };

Author: Oh I see. I'll update it then, but I believe the bloat point stands still. I think it'd be better to have a typed `resolvers.ts`.

>     };

Reviewer: No because it needs to infer the types for the `contextEnhancer`, having them in separate files it won't be achievable.

>     };

Reviewer 2: It should. That's an implementation detail. TS will infer it properly, no matter if it's from an import or not.

```ts
createGraphQLRuntime({ resolvers: { Query: { foo: () => "bar" }});
// is similar to
import resolvers from "./resolvers";
createGraphQLRuntime({ resolvers });
```

The reason why we recommend inline by default is because it's usually simpler to have everything in the same file, to follow the context of initialization and usage.

Because having small resolvers is a good practice, we saw that 80%+ of the time, both `resolvers` and `contextEnhancer ` fit in < 100 LOC. It's simpler to prevent an indirection in such cases.

But for big modules, keeping `resolvers` in a different file can make sense (even though in this case, it could be a smell of a module with too many responsibilities :angel:)

>     };

Author: That makes even more sense now, yes. Thanks for the clarification.

    ```
1.  Finally register the widget type from Magento and associate it with the
    GraphQL type

    ```ts title="extensions/acme-extension/modules/wysiwyg/runtime.ts"
      // ...
      contextEnhancer: ({ loaders }) => {
        loaders.MagentoWidget.registerWidgetType(
          // the tag name in your HTML
          "acme/product-preview",
          // The associated GraphQL type name
          "WidgetProductPreviewData"
        );
      },

>       // ...

Reviewer: Show the full example;
- missing the `loadRuntime` in the GraphqlModule
- missing the import and creatpr function for `createGraphQLRuntime({...})`

    ```

#### Get the data in your component

1.  Override the
    `theme/modules/Wysiwyg/MagentoWysiwyg/MagentoWysiwygFragment.gql` to fetch
    the new data needed for your widget
    ```graphql title="theme/modules/Wysiwyg/MagentoWysiwyg/MagentoWysiwygFragment.gql"
    fragment MagentoWysiwygFragment on MagentoWysiwyg {
      childNodes
      data {
        dataId
        ... on WysiwygWidgetData {
          data {
            ... on WidgetInvalidData {
              dataId
            }
            // highlight-start
            ... on WidgetProductPreviewData {
              product {
                sku
                name
              }
            }
            // highlight-end
          }
        }
      }
    }
    ```
1.  Use the fetched data in the `data` props in your final widget component (the

>     `./path/to/ProductPreview.js` mentioned above)

Reviewer: I can't find where the ProductPreview is mentioned?

It's not clear; 
- there is a `./path/to/ProductPreview.js` where is the `/path/to`? 
- is this a component, a function, a config?

Author: It was mentioned in the component lazyloading part. Now it's in the `lazy(() => import("theme/path/to/ProductPreview.js"))`
